Add nodes pre-provisioning

korablin opened the pull request:
In order to have instances at the ready at all times and
thus speed up CI runs because instance provisioning is done
upfront, add a mechanism and the related parameter to configure
a minimum number of instances that must be ready at all times.

Also add a mechanism to reconcile total number of nodes
based on template/cloud capacity

There is similar functionality presented in openstack cloud plugin:
https://github.com/jenkinsci/openstack-cloud-plugin/blob/master/plugin/src/main/java/jenkins/plugins/openstack/compute/JCloudsPreCreationThread.java

pjdarton commented:
Thanks for writing this - it's functionality that I approve of 👍 
I'm familiar with the openstack plugin and I like that feature 😉 

However, I'm not currently in a position to review this in detail right now, so I'll stick to general points:

`public` stuff: while this plugin code has loads more things public than it should, I don't want the problem made any worse. Please ensure that all externally-visible API additions are marked `@Restricted(NoExternalUse.class)` ... and use package access instead of public unless public is absolutely necessary.

WebUI: use help pages rather than description attributes. I find that description text bloats the (already long) page more than having a help icon on the right ... and this functionality could do with a big explanation of what it does so the help text would be a perfect place for that.

VSphereCloudRetentionStrategy.check might be better if it did it's thing and then, if the node isn't needed, delegated to super.check rather than copy/paste what super.check currently does ... or else, why not do what RunOnce does and call done(c)?  If there is a good reason why it has to be different, please comment why.

Unit tests would be appreciated; I know we can't test against a real vSphere instance, but we can mock things and check the code is requesting the changes we expect.
(And, yes, I admit that most of this code has no unit tests; I'd like to change that by having new functionality have better QA as this plugin is enough of a maintenance-nightmare as it is)

korablin commented:
Hi @pjdarton! 
Thanks for your fast and informative response. I have addressed your comments with the second commit except for the unit tests. I wouldn't mind to add them, but I don't see an easy way of mocking things that are needed to test this PR.
I would love to cover this code with tests once upstream will provide some guide/mechanism for easy testing.

Could you take a look once again at the code and see if something is still too 'public'? I am not very good at Java, so I could have missed a thing.

Thanks a lot in advance!

Regards,
Nikolai

pjdarton commented on src/main/resources/org/jenkinsci/plugins/vSphereCloudSlaveTemplate/config.jelly:
```diff
@@ -54,6 +54,10 @@
                 <f:textbox clazz="required number" default="0"/>
             </f:entry>
 
+            <f:entry title="${%Instances Min}" field="instancesMin" description="The number of instances to be pre-provisioned.">
```
We don't need/want a description when we have help text.

pjdarton commented on src/main/resources/org/jenkinsci/plugins/vSphereCloudSlaveTemplate/help-instancesMin.html:
```diff
@@ -0,0 +1,5 @@
+<div>
+    Number of instances to be provisioned beforehand.<br>
```
Can we have this of the form:

one line summary
<p>
Multiple sentences explaining what this does.
Possibly using <br> where appropriate.

If appreciate it if you can expand the description.
Tell folks what a value of 0 does.
Tell folks what a value of 1 does, and what happens when that one agent stays busy for ages.

... and my *guess* is that if you set it to the same value as the max instances then it'll try to keep that number provisioned regardless of Jenkins workload, replacing any VMs that close e.g. due to "run once" retention strategies... but if that's the case, please state this.
(Nobody ever complains about having too much help text, but too little often results in folks raising an issue asking WTF stuff does, and I'd prefer to avoid that)

pjdarton reviewed (commented):
Code looks good. All the newly-added public stuff now has `@Restricted` on it which is good.

WebUI resource files need tweaking.

Disclaimer: I've not tried this out myself yet; I'm just looking at the code changes on GitHub. I will (eventually) get around to trying it at work (and if it works, I'll make use of it right away!) but I'm not going to merge it until I've seen it working myself "for real".

pjdarton commented:
Note: if the Jenkins test fails and you don't think it's your fault, you should be able to get GitHub to re-run the test.
(and if GitHub doesn't let you do it easily, you can force it by closing the PR, waiting 10 minutes, then re-opening the same PR)

korablin commented:
Hey @pjdarton! Glad you are thrilled to get this functionality added. I agree that more docs is better than less.
What do you think about this one?
<img width="981" alt="Screenshot 2021-02-19 at 17 56 35" src="https://user-images.githubusercontent.com/13754260/108520930-43280580-72dc-11eb-9812-4faaafea5095.png">

korablin commented:
<img width="955" alt="Screenshot 2021-02-19 at 18 48 56" src="https://user-images.githubusercontent.com/13754260/108527272-26430080-72e3-11eb-89cf-442ccfe2f5fa.png">

korablin commented:
![Uploading Screenshot 2021-02-19 at 21.41.38.png…]()

korablin commented:
<img width="975" alt="Screenshot 2021-02-19 at 21 41 38" src="https://user-images.githubusercontent.com/13754260/108547403-7c23a280-72fb-11eb-9180-ea28deef3004.png">

korablin commented:
Hey @pjdarton! Hope you are doing well, mate.
Could you take a look at my last commit?

pjdarton reviewed (changes requested):
I think `Boolean` vs `boolean` is a simple coding error and easily fixed.

I suspect that many new methods are more visible than they need to be.  `public` should be the last-resort; default visibility is package-level and as most of the plugin's code is all in the same package that's often all that's needed. `protected` is generally only used to give access to subclasses.
...but we do need to ensure that all new non-private code is `@Restricted(NoExternalUse.class)` unless it's specifically part of the public API (so we don't restrict access to the new getter & setter fot the new field, but all the internal workings should be restricted)

I'm concerned that the "retain VM" functionality will actually defeat / prevent-from-working existing retention-strategy functionality that kills off VMs when they're no longer useful.
It may be better to have this functionality create new VMs to fulfil the instanceMin requirement but allow them to time out normally, instructing users 9in the documentation) that if they don't want instances to time out quickly like that then they'll need to increase the retention time.
The existing "if instancesPresent < instanceMin then ignore the retention strategy and keep it" logic check seems to be _too_ simplistic.

Dealing with the asynchronicity of retention-strategy checks and the instanceMin provisioning background task _appears_ to be improperly handled.  I suspect that you'll need some more complicated state handling for this to work properly.
Consider having the periodic task evaluate which instances ought to be kept, and in what priority order, and have the retention strategies then act on that evaluation result "when they get called" ... or (and I suspect this will be the much simpler option) have the periodic task _only_ be responsible for creating new instances if the minimum isn't met yet and let the retention strategies take care of disposing of them.

korablin commented:
@pjdarton I have killed `shouldSlaveBeRetained()` function and removed changes that I did to ret policies, so they their work is not anymore affected by this change.
Initially, I added those changes cause I wanted to make this behaviour consistent with OpenStack plugin, but guess we could iterate on this later if people will be complaining.

Regarding your question on using package-level visibility: I don't think it is possible.
Methods that reconcile scripts uses are located in `package org.jenkinsci.plugins`, but the script itself is in `package org.jenkinsci.plugins.vsphere` which appear to me as different packages.
Correct me if I am wrong. Also I am not sure if this the right place for this script.

korablin commented:
I am sorry for the belated response.
It would be nice to communicate quickly. I am willing to address your future comments asap.

pjdarton commented on src/main/java/org/jenkinsci/plugins/vsphere/VSphereNodeReconcileWork.java:
```diff
@@ -0,0 +1,91 @@
+                int maxNodes = Math.min(globalMaxNodes, templateMaxNodes);
+
+                // if maxNumber is lower than instancesMin, we have to ignore instancesMin
+                int toProvision = Math.min(instancesMin - idleNodes.size(), maxNodes - runningNodes.size());
```
I'm not convinced that it's as simple as this as there's a delay between when a provisioning is started and when Jenkins is given a new agent and hence creates a new `vSphereCloudSlaveComputer` instance (and again when it's time to delete a VM too).

If you check the logic in vSphereCloud.provision(...) you'll see that it calls `ensureLists()` and then interrogates its `templateState` data to count how many VMs exist.
Some of those VMs will be in a state where the vSphere service has been asked to create a new VM but Jenkins won't know it exists yet; some of them will be fully created VMs that aren't online yet; some will be online and either busy or idle; some will be in the process of being deleted.
That `templateState` data structure is the "one true source of information" that we have (in memory) of what VMs we've asked for, what we've got, and what we've asked to delete.

The logic in use here only considers VMs that are fully created and online (and the subset of those which are idle); it does not consider those "not yet online" or "those being deleted".

I think that what we need here is to use the same logic (or at least be consistent with it) as used in vSphereCloud so that VMs which are being created are counted against the total present in the cloud, otherwise we could easily overshoot the desired `instancesMin` value (e.g. if VMs take so long to create that `public void execute(TaskListener listener)` gets called again before instances it's already requested have "arrived".
IMO this "pre-create" functionality will be most valued by folks whose vSphere infrastructure is slow to create new VMs, so this code _must_ cope with the infrastructure being slow.

pjdarton commented on src/main/java/org/jenkinsci/plugins/vsphere/VSphereNodeReconcileWork.java:
```diff
@@ -0,0 +1,91 @@
+                        LOGGER.log(Level.SEVERE, "Failed to pre-create instance from template {0}. Exception: {1}",
+                                new Object[] { templateLabel, ex });
+                    }
+                } else if (toProvision < 0) {
```
I am not sure we need this kind of logic here - the retention policy should allow unused and unnecessary nodes to be quietly disposed of.
I'm open to discussion on this, but e.g. it isn't deemed necessary in the OpenStack plugin and I suspect that its implementation works fine.

pjdarton commented on src/main/java/org/jenkinsci/plugins/vsphere/VSphereNodeReconcileWork.java:
```diff
@@ -0,0 +1,91 @@
+                } else if (toProvision < 0) {
+                    int toDelete = Math.min(idleNodes.size(), Math.abs(toProvision));
+                    for (int i = 0; i < toDelete; i++) {
+                        AbstractCloudSlave node = idleNodes.get(i).getNode();
```
There's a potential race condition here:  By the time we get here, agents may not be idle anymore (and may not even exist within Jenkins anymore).
(Jenkins is _heavily_ multi-threaded - loads of things are happening all at once)

pjdarton commented on src/main/java/org/jenkinsci/plugins/vsphere/VSphereNodeReconcileWork.java:
```diff
@@ -0,0 +1,91 @@
+                    LOGGER.log(Level.INFO, "Pre-creating {0} instance(s) for template {1} in cloud {3}",
+                            new Object[] { toProvision, templateLabel, cloud.name });
+                    try {
+                        cloud.provision(label, toProvision);
```
The problem here is that `label` isn't neccessarily unique to one template.

What we want here is to provision an instance for a specific template, rather than leaving it to the `provision` method to select a template that matches the specified label.
(FYI the `provision` method will distribute VMs over all templates matching the specified label if those templates have instance caps on them)

e.g. where I work, we have set of GPU-enabled servers on our vSphere cluster, each of which is capable of running N VMs, so we have one template per server with "max instances" = N, but aside from a specification to limit the template to a specific server, the templates are identical.  This means that if we're empty and then get a request for 5 instances that can run on either or two servers, we'll run 3 on one server and 2 on the other.
...although, in this example, all 5 would take time in arriving (as our vSphere cluster takes a while to clone a new VM from a vSphere template) and then take even longer before they're online (as the VMs run Windows and this takes a while to boot up).

If you compare what you're doing here with what was done on the [OpenStack PR](https://github.com/jenkinsci/openstack-cloud-plugin/pull/205/files) you'll see that the provisioning process in the cloud class was refactored to expose a "provision a new agent from this template" method; I'd suggest that we're going to need a similar thing here, otherwise this method will confuse templates and labels.

korablin commented:
@pjdarton  thanks for another iteration on this PR. I appreciate it.
So far, I pushed the minor changes fixing level of visibility (those things used to be public due to earlier functionality that I've killed).

Regarding the major things:
   - Tracking state of a request i.e. using templateState when deploying a VM + provisioning templates instead of labels
This sounds right to me. Could you give a hint on how do you envision implementation for this?
Should it be a method similar to `provision` in vSphereCloud that takes `template` and number of instances as arguments and then operates with `CloudProvisioningRecord` ? 

I would love to hear your opinion before I will start implementing it.


Also, current combination of provisioning things the simple way (as I do now) + deletion of redundant nodes within this code, kind of makes sure that we meet the instancesMin number, but of course there is a risk creating excessive VMs that will be killed on the second run of the periodic job.
^^ I am mentioning this just in case, if we would consider merging the code as is and iterate on it later with further improvements - that shouldn't blow up  your Jenkins :)

pjdarton commented:
Re: hint on how do you envision the implementation

A cloud plugin's method `public Collection<PlannedNode> provision(final Label label, int excessWorkload)` is how Jenkins asks the cloud-plugin to provide more nodes; this method choses a template to use and then creates a new node (and then repeats until enough executors are in progress, finally returning a set of "planned nodes" to Jenkins).
Within that code, it maintains the internal records (`templateState.pruneUnwantedRecords()`), ensures it doesn't exceed the cloud capacity (and returns `Collections.emptySet()` if so) before getting on with deciding which template(s) match the label Jenkins is asking for and then creating new VMs from the template it decided on.

This PR is adding a new method that'll provide more nodes; it'll need to iterate over all the templates checking that there's enough <existing idle nodes> + <nodes under construction> to satisfy the template's instancesMin field.


In this new case, we don't want to call `List<vSphereCloudSlaveTemplate> templates = getTemplates(label)` as, instead, we already know exactly which template we want (as it's iterating over all of them) ... but we could just make a list containing just the template we want before going on to call `templateState.calculateProvisionableTemplates(templates)` etc.

So, I think that what we do is

1. Refactor the existing vSphereCloud `provision` method so that _some_ of what it does is split out into separate methods (that our new code can call).  [This bit](https://github.com/jenkinsci/vsphere-cloud-plugin/pull/127/files#diff-4f085b36000ec240357f1171a2a163cad43c64c3e1bafacc8a60804007271cbdR374-R379) looks promising, but there may be others too.
2. Define a new `provisionInstanceMin` method in vSphereCloud that'll do `synchronized (this) { ensureLists(); }` and then, within a `synchronized (templateState)` block, will iterate over each of the `CloudProvisioningRecord`s in `templateState`, compare the `instanceMin` field's value with the number of `CloudProvisioningRecord.size()` (as that'll include all VMs, not just active nodes) minus the number of _active_ nodes for this template and provision some new ones if more are needed (and there's space for them!).  There's a fair amount of decision making etc that'll be identical to that done in the original `provision` method.
3. I would suggest you remove the existing code that'll dispose of unwanted instances - disposal is hard to do safely (e.g. as shown in [JENKINS-55066](https://issues.jenkins.io/browse/JENKINS-55066) and similar) and there are existing mechanisms that will slowly dispose of unwanted VMs safely.
4. have the periodic code loop over the vSphereClouds and, for each of those, call the new provisionInstanceMin method.
5. Finally, compare the new `provisionInstanceMin` method with the old vSphereCloud `provision` method and check again for common code that ought to be refactored into its own method that's then used by both.

...and be careful not to change the behaviour of the original `provision` method - no matter how much the code is rearranged, the functionality of that method needs to remain unchanged.

korablin commented:
@pjdarton It's been a while... I am sorry again. Ever-changing priorities are running plans sometimes.
Hopefully, with the last commit that refactors the whole thing for the better way, we will be able to merge it soon :)
Thanks for all of your comments! I have tried my best to address them and things look much simpler now.

I am willing to address further comments asap. 
Thank you @pjdarton in advance!

korablin commented on src/main/java/org/jenkinsci/plugins/vSphereCloud.java:
```diff
@@ -352,22 +352,15 @@ private Integer calculateMaxAdditionalSlavesPermitted() {
-                        }
-                        maxSlavesToProvisionBeforeCloudCapHit = Integer.valueOf(intValue - 1);
-                    }
+                    if (!cloudHasCapacity()) break;
```
I am not subtracting 1 here, since `maxSlavesToProvisionBeforeCloudCapHit` is re-calculating on every call to `cloudHasCapacity()`

pjdarton commented on src/main/java/org/jenkinsci/plugins/vsphere/tools/CloudProvisioningAlgorithm.java:
```diff
@@ -81,6 +81,23 @@ public static String findUnusedName(CloudProvisioningRecord record) {
+        int provisionedNodes = record.getCurrentlyProvisioned().size() + record.getCurrentlyPlanned().size();
+        int requiredPreProvisionedNodes = record.getTemplate().getInstancesMin();
+        return requiredPreProvisionedNodes - provisionedNodes;
+    }
```
We're going to need a unit test for this new method.

pjdarton commented on src/main/java/org/jenkinsci/plugins/vsphere/VSpherePreProvisonWork.java:
```diff
@@ -0,0 +1,52 @@
+    public void execute(TaskListener listener) {
+        for (Cloud cloud : Jenkins.getActiveInstance().clouds) {
+            if (!(cloud instanceof vSphereCloud)) continue;
+            for (vSphereCloudSlaveTemplate template : ((vSphereCloud) cloud).getTemplates()) {
```
Rather than casting `((vSphereCloud)cloud)` every time, cast once into a variable and use that.
e.g.
```
if (!(cloud instanceof vSphereCloud)) continue;
vSphereCloud vsCloud = (vSphereCloud) cloud;
for (vSphereCloudSlaveTemplate template : vsCloud .getTemplates()) {
```
Casts are ugly - they're best done as little as possible 😉

pjdarton reviewed (changes requested):
We need a unit-test for the new method that's been added to CloudProvisioningAlgorithm (that code can be tested in isolation and has good coverage)

I would _like_ to have a unit-test for the new provisoning code too - some test scenarios where the auto-provisioning is called (I think we can assume Jenkins will call it and don't need to test that Jenkins does so) with different situatuons and check that it decides (correctly) what to provision (stubbing/mocking out the bits that'd talk to a vSphere instance).
e.g. "if we've got no instances, no instance limit and instanceMin is 2 then we expect two instances to be provisioned" and "if we've got 2 instances, no instance limit, and instanceMin is 2 then we expect none to be provisioned" etc.
IME unit-tests are a good way of flushing out "out by 1" mistakes (that are hard to spot "in the real world") and they also help protect functionality from getting broken by later changes (changes done by people who don't use this feature and hence won't notice if they break it _unless_ there's a unit test that tells them)

korablin commented on src/test/java/org/jenkinsci/plugins/NodePreProvisionTest.java:
```diff
@@ -0,0 +1,109 @@
+        assertThat(stubVSphereTemplateState.countNodes(), equalTo(1));
+
+        // Here it says that there still should be provisioned 2 nodes, despite the fact there is 1 active already
+        stubVSphereCloud.preProvisionNodes(template);
```
The logs: 
```
Jun 01, 2021 5:54:23 PM org.jenkinsci.plugins.vSphereCloud InternalLog
INFO: STARTING VSPHERE CLOUD

Jun 01, 2021 5:54:23 PM org.jenkinsci.plugins.vsphere.tools.CloudProvisioningState logStateChange
WARNING: Marking templateCapacity10instanceMin21 as active : wasPreviouslyPlanned!=true

Jun 01, 2021 5:54:24 PM org.jenkinsci.plugins.vSphereCloud preProvisionNodes
INFO: preProvisionNodesForTemplate(null): should pre-provision 2 nodes

Jun 01, 2021 5:54:24 PM org.jenkinsci.plugins.vSphereCloud calculateMaxAdditionalSlavesPermitted
INFO: There are 0 VMs in this cloud. The instance cap for the cloud is 100, so we have room for more

Jun 01, 2021 5:54:24 PM org.jenkinsci.plugins.vSphereCloud calculateMaxAdditionalSlavesPermitted
INFO: There are 1 VMs in this cloud. The instance cap for the cloud is 100, so we have room for more
```

korablin replied:
Note,  that when it preProvision is invoked it actually sees that "There are 1 VMs in this cloud", which was "provisioned" by the method just a moment ago.
So, I would love to achieve the same state without calling preProvision(); I've tried marking the associated record as `provisionedSlaveNowActive` or `provisioningStarted`, but that doesn't do the trick, neither calling `VSpherePlannedNode.createInstance(stubVSphereTemplateState, nodeName, provisionable);`

korablin commented:
Hey @pjdarton ! Thanks for the comments :) 
I've added a draft code with unit tests for the new provisioning method: "NodePreProvisionTest.java"
But I can't get it working properly, though. Could you help me to understand what is wrong there?

I couldn't find a way to convince `preProvison` method that there is something provisoned already.

korablin commented:
Hey @pjdarton! If I remove "NodePreProvisionTest.java" do you think this could be landed?
